src/parser/shared/normalizers/Channeling.ts

```diff
@@ -0,0 +1,344 @@
+    buffChannelSpec(SPELLS.CRACKLING_JADE_LIGHTNING.id),
+    buffChannelSpec(SPELLS.FISTS_OF_FURY_CAST.id),
+    // Demon Hunter
+    buffChannelSpec(SPELLS.EYE_BEAM.id), // TODO special handling because of the two buffs?
+    // Shaman
+    // Hunter
+    buffChannelSpec(SPELLS.RAPID_FIRE.id),
+    // TODO Barrage - can't find a single log with someone taking the talent...
+    // Paladin
+    // Warrior
+    buffChannelSpec(SPELLS.BLADESTORM_TALENT.id),
+    buffChannelSpec(SPELLS.BLADESTORM.id),
+    // Death Knight
+    buffChannelSpec(SPELLS.BLOODDRINKER_TALENT.id),
+  ];
+
+  // registered special case handlers, mapped by guid
+  channelSpecMap: { [key: number]: ChannelHandler } = {};
+
+  /**
+   * Constructs a Channeling normalizer that deals with normal hardcasts and instants by default,
+   * and then allows special case functions to be registered by spellId for channels.
+   * @param channelSpecs special case functions that handle specific spells
+   */
+  constructor(options: Options) {
+    super(options);
+
+    // populate the specs into a mapping for quick handling
+    Channeling.CHANNEL_SPECS.forEach((spec) => {
+      spec.guids.forEach((guid) => {
+        if (this.channelSpecMap[guid]) {
+          console.error(
+            'Channeling module trying to add more than one handler for spell guid ' +
+              guid +
+              ' - only the latest handler will apply!',
+          );
+        }
+        this.channelSpecMap[guid] = spec.handler;
+      });
+    });
+  }
+
+  normalize(events: AnyEvent[]) {
+    const eventsInserter: InsertableEventsWrapper = new InsertableEventsWrapper(events);
+
+    const channelState: ChannelState = {
+      unresolvedChannel: null,
+      eventsInserter,
+    };
+
+    // for each event, check if there is a special case handler for that GUID -
+    // if so - call the handler, if not - call the default handling
+    events.forEach((event: AnyEvent, index: number) => {
+      if (!HasSource(event) || event.sourceID !== this.selectedCombatant.id) {
+        return; // only concerned with events coming from selected player
+      }
+      const handler = HasAbility(event) ? this.channelSpecMap[event.ability.guid] : undefined;
+      if (handler) {
+        handler(event, events, index, channelState);
+      } else {
+        this.defaultHandler(event, events, index, channelState);
+      }
+    });
+
+    return eventsInserter.build();
+  }
+
+  /**
+   * Handles the default cases of regular cast time spells, instants, and 'fake' casts
+   */
+  defaultHandler(
+    event: AnyEvent,
+    events: AnyEvent[],
+    eventIndex: number,
+    channelState: ChannelState,
+  ): void {
+    if (event.type === EventType.BeginCast) {
+      // we're starting a new generic hardcast
+      cancelCurrentChannel(event, channelState);
+      beginCurrentChannel(event, channelState);
+    } else if (event.type === EventType.Cast) {
+      // TODO this won't catch pre-cast channels
```

Comment:
is it even possible to precast a channel? like, won't the first application initiate combat? i guess its a question of whether that hit gets logged which...maybe?

Reply:
There are some pets that have channels like Crane clones for all monks, yu'lon the jade serpent mistweaver
And i'd imagine there are some for demo and maybe water elemental
They all won't affect downtime calculation but it could yield some useful information about the pet AI

Reply:
Yes it is possible to precast a channel and more logs exist than you would assume with it. 
Mistweavers with essence font and marksmen hunters with aimed shot

Reply:
This is in the defaultHandler, the comment is really referring to hardcasts. Basically, if a player starts a cast before the pull but finishes it after, we'll see the cast event but not the begincast, and so this handler will believe it's an instant and handle it accordingly.

This will cause only small inaccuracy, and the handling (and comment) are carryovers from the previous handler, so my new code is at least not any worse.